binary-data-formats.qmd

```diff
@@ -0,0 +1,404 @@
+To create a multi-file data set with one parquet file per year of library data we would do the following:
+
+```{r write-dataset}
+#| eval: !expr "!file.exists('data/seattle-library-checkouts/CheckoutYear=2021/part-0.parquet')"
+seattle_lib |>
+  group_by(CheckoutYear) |>
+  write_dataset(
+    path = "data/seattle-library-checkouts",
+    format = "parquet"
+  )
+```
+
+Let's take a look at the files:
+
+```{r show-parquet-files}
+seattle_lib_files <- list.files(
+  path = "data/seattle-library-checkouts",
+  recursive = TRUE
+) 
+seattle_lib_files
+```
+
+These filenames follow a "self-describing" convention used by the Apache Hive project ([hive.apache.org](https://hive.apache.org/)).
+Hive-style partitions name folders using a "key=value" convention, and the data files in that folder contain the subset of the data for which the key has the relevant value.
+Following this convention makes it easier to locate files relevant to a particular query.
+
+### Performance advantages
+
+Partitioning a large data set into many smaller, meaningful files can produce substantial improvements in performance.
+To illustrate, we'll repeat the analysis of the 2021 Seattle library book checkouts by month using the partitioned data:
+
+```{r dataset-performance-multiple-parquet}
+#| cache: true
+system.time({
+  open_dataset(
+    sources = "data/seattle-library-checkouts/", 
+    format = "parquet"
+  ) |>
+    filter(CheckoutYear == 2021, MaterialType == "BOOK") |>
+    group_by(CheckoutMonth) |>
+    summarise(TotalCheckouts = sum(Checkouts)) |>
+    arrange(desc(CheckoutMonth)) |>
+    collect()
+})
+```
+
+This is a 10-fold speed up compared to a single parquet file, and a 100-fold speed up compared to the original csv file.
+
+## Other topics
+
+I haven't quite worked out the best way to integrate these sections.
+
+### Using duckdb with arrow
+
+One useful feature of arrow is that it is interoperable with duckdb and dbplyr.
+It is possible to open a multi-file data set using arrow and then pass it to duckdb using the `to_duckdb()` function.
+Doing so means that your dplyr code will now be interpreted by dbplyr (rather than arrow) and executed by duckdb.
+
+```{r use-duckdb}
+open_dataset("data/seattle-library-checkouts/") |>
+  to_duckdb() |>
+  filter(CheckoutYear == 2021, MaterialType == "BOOK") |>
+  group_by(CheckoutMonth) |>
+  summarise(TotalCheckouts = sum(Checkouts)) |>
+  arrange(desc(CheckoutMonth)) |>
+  collect()
+```
+
+There is an analogous function `to_arrow()` that passes control back to arrow, meaning that it is possible to swap back and forth between the duckdb and arrow back ends within a single analysis pipeline.
+This operation is extremely efficient (only the metadata need to be copied from one system to the other) can occasionally be valuable when one system implements functionality that the other does not.
```

Comment:
```suggestion
This operation is extremely efficient because both libraries deal with the Arrow format, so data is passed between then without making a copy. This can occasionally be valuable when one tool implements functionality that the other does not.
```

Reply:
If you take Neal's suggestion of 

```
This operation is extremely efficient because both libraries deal with the Arrow format, so data is passed between then without making a copy. This can occasionally be valuable when one tool implements functionality that the other does not.
```

We might also add in something about serialization as well as the copying bits. Both are important, but in some ways not needing to (re)encode the data to send it is the biggest time saver